Add basic build and test

ccremer opened the pull request:
## Summary

* Adds `Makefile` and `Makefile.vars.mk`
* Adds GitHub workflows for lint, build and tests
* configures CodeClimate

## Checklist

- [x] Categorize the PR by setting a good title and adding one of the labels:
      `bug`, `enhancement`, `documentation`, `change`, `breaking`, `dependency`
      as they show up in the changelog

<!--
Remove items that do not apply. For completed items, change [ ] to [x].

NOTE: these things are not required to open a PR and can be done afterwards,
while the PR is open.
-->

ccremer committed:
Add Makefile for build and test

ccremer committed:
Add test workflow

ccremer committed:
Add sample test case

ccremer committed:
Add lint workflow

ccremer committed:
update todo's

glrf commented on Makefile:
```diff
@@ -0,0 +1,50 @@
+
+.PHONY: test.go
+test.go: ## Run unit tests against code
+	go test -race -coverprofile cover.out -covermode atomic -count 1 ./...
```
Why do you force the retesting? (with `-count 1`) This can be helpful if the tests are non deterministic but I think for the vast majority of the applications I would prefer to keep test caching.

ccremer replied:
I've come across a few cases in the past where the test cache was actually too aggressive and didn't execute test cases even after code changes, thus falsifying the results. Those were very hard to find out and debug. That's why I often "disable" that cache. However, maybe it was caused by an older Go version and fixed in the meantime :shrug:

glrf commented on Makefile:
```diff
@@ -0,0 +1,50 @@
+build.docker: build.bin ## Build docker image
+	docker build -t $(CONTAINER_IMG) .
+
+.PHONY: test
+test: test.go ## All-in-one test
+
+.PHONY: test.go
+test.go: ## Run unit tests against code
```
Very nitpicky so feel free to ignore this,  but I was a little confused by the `test.go` target because it looks so much like a file. I would personally prefer a `test-go` (and `build-bin` etc.) just to make it clear that we are not using file targets. This is also generally more in line with [GNU's naming convention](https://www.gnu.org/software/make/manual/html_node/Standard-Targets.html#Standard-Targets)(although there does not seem to be an actual style guide so :shrug:)

ccremer replied:
So the initial motivation comes from a syntax that I kinda like which I''ve seen in Laravel. Hierarchical targets like
```
build
build:bin
build:docker
...
```
However, make doesn't like the `:` as a separator, unless it's escaped with `\:`, but this causes issues like broken shell autocomplete when doing `make <tab>` and also the Makefile becomes way more ugly.
So I thought about just the single dot `.`, but I see the argument that it looks like a filename.
ofc dash would work also, but I see it more like a word separator, less as a hierarchy separator. But I guess that's the `make` way, so I'll change to dash as a separator unless you have other ideas :)

bastjan commented on Makefile:
```diff
@@ -0,0 +1,50 @@
+	git diff --exit-code
+
+.PHONY: generate
+generate: ## Generate additional code and artifacts
```
I think we could leave `go generate ./...` in. It's a no-op if there's nothing to generate.

ccremer committed:
Improve makefile

Get rid of dot as a hierarchy separator, as they might confuse to be file names.
